### DISCUSSION TITLE
Passing a StringBuilder to an interpolated string

### DISCUSSION THREAD

**Discussion by friflo** [dotnet/csharplang#7336] [Language Ideas] [answered]
Hi,
did not find a similar open or closed issue.

A typical use case is the creation of SQL statements. Code typically looks like this
```csharp
    var sql = new StringBuilder();
    sql.Append($"INSERT INTO {name} ({col1},{col2},{col3}, ...)\nVALUES ");
    AppendRecords(sql, records);
```
So you have both **string interpolation** and appending strings or scalars manual to a `StringBuilder` in `AppendRecords()`.
The **string interpolation** creates its own `StringBuilder` instance internally.
Both instances will create (to resize) `char[]` instances if required.

The idea is simply to enable passing the `StringBuilder` instance to the interpolated string.
It could look like this:
```csharp
    var sql = new StringBuilder();
    $(sql)"INSERT INTO {name} ({col1},{col2},{col3}, ...)\nVALUES ";
    AppendRecords(sql, records);
```
As a result:
- less code
- less object instantiations.
  - -1 `StringBuilder` instance
  - less `char[]` instances

**Comment by huoyaoyuan** [comment]
> A typical use case is the creation of SQL statements.

**DO NOT use string builder or raw string manipulation to construct SQL**. It's the most common source of SQL injection risk.

> The **string interpolation** creates its own `StringBuilder` instance internally.

It doesn't. From .NET 6 and C# 10, the interpolated string won't be constructed into a string first. It's passed as raw parts to an interpolated string handler defined by the callee, if there is one. See https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/interpolated-string-handler.

**Reply by HaloFour** [reply, marked as answer]
And `StringBuilder` does explicitly have an overload of `Append` which supports an interpolation handler, so:

```cs
var sql = new StringBuilder();
sql.Append($"INSERT INTO {name} ({col1},{col2},{col3}, ...)\nVALUES ");
```

is lowered into:

```cs
StringBuilder sql = new StringBuilder();
StringBuilder.AppendInterpolatedStringHandler handler = new StringBuilder.AppendInterpolatedStringHandler(30, 4, sql);
handler.AppendLiteral("INSERT INTO ");
handler.AppendFormatted(value);
handler.AppendLiteral(" (");
handler.AppendFormatted(value2);
handler.AppendLiteral(",");
handler.AppendFormatted(value3);
handler.AppendLiteral(",");
handler.AppendFormatted(value4);
handler.AppendLiteral(", ...)\nVALUES ");
sql.Append(ref handler);
```

Where the `StringBuilder.AppendInterpolatedStringHandler` is a `struct` that just delegates the calls back to the original `StringBuilder`.

**Reply by friflo** [reply]
Okay thx, Was not aware of this.
Creating an SQL statement was for illustration - could have been another use-case.

**Reply by alex-jitbit** [reply]
@HaloFour Wow, this is great and I've actually verified this via SharpLab here: https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU+hxTAbgIYBO6AzgEboC86AHYBTAO6Y4AOgAqIgB4AXKQGVF7AJZCA5gCEArhoA2wEezoBuRkyJbF6VgPRwYqK0hvFeUgIIAHPxEhYGoAEgAiLmAAMyi8VgBfcNp3JgTkBKA===

But which .NET runtime uses this? Looks like .NET 6 and above - correct?

**Reply by 333fred** [reply]
Yup. You can read all about it (in lots of detail) in Stephen Toub's blog on the subject here: https://devblogs.microsoft.com/dotnet/string-interpolation-in-c-10-and-net-6/.
